ccf_specs.md

```diff
@@ -161,6 +161,91 @@ This specification uses CDDL notation to express CBOR data items:
+
+Decoders MUST reject CCF encodings that are not valid unless otherwise specified in this section.
+
+A CCF encoding complies with "Valid CCF Encoding Requirements" if it complies with the following restrictions:
+
+- CBOR data items MUST be well-formed and valid as defined in RFC 8949.  E.g. CBOR text strings MUST contain valid UTF-8.  As an exception, RFC 8949 requirements for CBOR maps are not applicable because CCF does not use CBOR maps.
+
+- CCF encodings must comply with specifications in "CCF Specified in CDDL Notation" section of this document.
+
+- `composite-type.id` MUST be unique in `ccf-typedef-message` or `ccf-typedef-and-value-message`.
+
+- `composite-type.cadence-type-id` MUST be unique in `ccf-typedef-message` or `ccf-typedef-and-value-message`.
+
+- `field-name` MUST be unique in `composite-type`.
+
+- `type-ref.id` MUST refer to `composite-type.id`.
+
+- `composite-type-value.id` MUST be unique in the same `composite-type-value` data item.
+
+- `type-value-ref.id` MUST refer to `composite-type-value.id` in the same `composite-type-value` data item.
+
+- all parameter lists MUST be unique in `composite-type-value`.
+  - `name` MUST be unique in `composite-type-value.fields`. 
+  - `identifier` MUST be unique in `composite-type-value.initializers`.
+
+- elements MUST be unique in `restricted-type` or `restricted-type-value`.
+
+- keys MUST be unique in `dict-value`.  Decoders are not always required to check for duplicate dictionary keys.  In some cases, checking for duplicate dictionary key is not necessary or it may be delegated to the application.
+
+### Deterministic CCF Encoding Requirements
+
+A CCF encoding is deterministic if it satisfies the "Deterministic CCF Encoding Requirements".
+
+Encoders SHOULD emit CCF encodings that are deterministic.  CCF-based protocols MUST specify when encoders are required to emit deterministic CCF encodings.
+
+Decoders SHOULD check CCF encodings to determine whether they are deterministic encodings.  CCF-based protocols MUST specify when decoders are required to check for deterministic encodings and how to handle nondeterministic encodings.
+
+A CCF encoding satisfies the "Deterministic CCF Encoding Requirements" if it satisfies the following restrictions:
+
+- CCF encodings MUST satisfy "Valid CCF Encoding Requirements" defined in this document.
+
+- CCF encodings MUST satisfy "Core Deterministic Encoding Requirements" defined in RFC 8948 Section 4.2.1.  As an exception, RFC 8949 requirements for CBOR maps are not applicable because CCF does not use CBOR maps.
+
+- `composite-type.id` in `ccf-typedef-and-value-message` MUST be identical to its zero-based index in `composite-typedef`.
+
+- `composite-type-value.id` MUST be identical to the zero-based encoding order `type-value`.
+
+- `inline-type-and-value` MUST NOT be used when type can be inferred.
+
+- The following data items MUST be sorted using bytewise lexicographic order of their deterministic encodings:
```

Comment:
Maybe describe below, shortly, on a very high-level, how an encoder could do this, e.g.
1. Gather all types from all values
2. Sort the Cadence type IDs of the types, assign CCF type IDs (implicitly by index) 
3. Encode the types (i.e. the information; this can now use the CCF type IDs)

Reply:
Yeah, that is close enough to what I described during our CCF specs meeting yesterday.

This can be done in a separate PR.  I can add a new section with a summary that more closely matches the CCF codec I've been developing.  It already has this feature so the summary can be a bit more detailed if needed.

Reply:
Yep, let's do it in a separate PR. Nice to have, but not important